### PR TITLE
Refactor and cleanup of the elf plugin

### PR DESCRIPTION
**Pull request by ghost**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This PR is huge and will be painful to review, so before that you can listen to a nice song: https://soundcloud.com/powerwolf-official/beast-of-gevaudan

I split elf.c in multiple sub file (Those files can be merged later)

Steps (I tried to do one commit for each step but I believed they should be merged):
* Move the function out of the file.
* Add documentation (I forget to document some function, and I am not satisfied with the documentation I add. So if you have any suggestion...)
* Refactoring (I dropped the refactoring of rz_bin_elf_get_symbols_with_type because I spend to much time on this. (3 days of my life, that feel like hell). I will try to rewrite the function from scratch after the GSoC)

I modified some bad test.

Sources: https://hackmd.io/@oszj3t95T6m-bk8YPxSjhA/Byf0LpXYd
Imported version of https://github.com/rizinorg/rizin/pull/1156

**Test plan**
```
rz-test -t 10
```

**Closing issues**
...

### CODE REVIEW

**Comment by XVilka** [comment]
Quite impressive job!

I suggest to split the code into the categories instead:
- `elf_sections.c`
- `elf_symbols.c`
- `elf_segments.c`
- `elf_information` - where all these `is_executable`, `is_nx`, etc go
- `elf_relocs.c`

File: librz/bin/format/elf/rz_bin_elf_deprecated.inc [hunk 1]
```diff
@@ -188,6 +189,7 @@ ut64 Elf_(rz_bin_elf_resize_section)(RzBinFile *bf, const char *name, ut64 size)
 	return delta;
 }
 
+// DEPRECATED FUNCTION
 /* XXX Endianness? */
 bool Elf_(rz_bin_elf_del_rpath)(RzBinFile *bf) {
```
Comment (thestr4ng3r) [inline librz/bin/format/elf/rz_bin_elf_deprecated.inc]: You can prefix with `RZ_DEPRECATE` for such cases.

**Review by thestr4ng3r — COMMENTED** [review]
I agree, I think such categories or similar can make sense. Then we could also compile them as individual objects instead of includes of `.inc`.

Apart from that, could you pleaye write down any changes you made that are not pure refactoring/moving and might affect behavior? I see for example there are some new members in `Elf_(rz_bin_elf_dynamic_info)`.

**Comment by thestr4ng3r** [comment]
`elf_core.c` or `elf_notes.c` might also be a category worth having.

**Comment by ghost** [comment]
* librz/bin/format/elf/rz_bin_elf_get_number_of_dynamic_symbols.inc (compute the number of dynamic symbols)
* Modified almost everything related to version info (librz/bin/format/elf/rz_bin_elf_get_version_info.inc)
* Heavy refactoring on librz/bin/format/elf/rz_bin_elf_get_symbols_from_phdr.inc
* librz/bin/format/elf/rz_bin_elf_get_sections.inc (Fixed plt section name)

**Comment by ghost** [comment]
```
# elf_sections.c

RZ_IPI bool Elf_(rz_bin_elf_is_sh_index_valid)(RZ_NONNULL ELFOBJ *bin, Elf_(Half) index);

RZ_BORROW RzBinElfSection *Elf_(rz_bin_elf_get_section)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL const char *section_name);
RZ_OWN char *Elf_(section_type_to_string)(ut64 type);
ut64 Elf_(rz_bin_elf_get_section_addr)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL const char *section_name);
ut64 Elf_(rz_bin_elf_get_section_addr_end)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL const char *section_name);
ut64 Elf_(rz_bin_elf_get_section_offset)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL const char *section_name);

# elf_symbols.c

RZ_IPI Elf_(Sym) Elf_(rz_bin_elf_get_symbol_entry)(RZ_NONNULL ELFOBJ *bin, ut64 offset);
RZ_IPI RZ_BORROW RzBinElfSymbol *Elf_(rz_bin_elf_get_symbols_from_phdr)(RZ_NONNULL ELFOBJ *bin, int type);
RZ_IPI RZ_BORROW const char *Elf_(rz_bin_elf_symbol_bind_to_str)(RZ_NONNULL Elf_(Sym) * sym);
RZ_IPI RZ_BORROW const char *Elf_(rz_bin_elf_symbol_type_to_str)(ELFOBJ *bin, RzBinElfSymbol *ret, RZ_NONNULL Elf_(Sym) * sym);
RZ_IPI RZ_OWN RzBinElfSymbol *Elf_(rz_bin_elf_compute_symbols_from_phdr)(RZ_NONNULL ELFOBJ *bin, int type);
RZ_IPI RZ_OWN RzVector *Elf_(rz_bin_elf_compute_symbols_from_segment)(RZ_NONNULL ELFOBJ *bin, int type, ut64 offset, size_t num, ut64 entry_size);
RZ_IPI RzBinElfSymbol *Elf_(rz_bin_elf_get_symbols_with_type)(ELFOBJ *bin, int type);
RZ_IPI size_t Elf_(rz_bin_elf_get_number_of_dynamic_symbols)(RZ_NONNULL ELFOBJ *bin);
RZ_IPI ut64 Elf_(rz_bin_elf_get_import_addr)(RZ_NONNULL ELFOBJ *bin, int symbol);
RZ_IPI void Elf_(rz_bin_elf_set_import_by_ord)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL RzBinElfSymbol *symbol);

RZ_BORROW RzBinElfSymbol *Elf_(rz_bin_elf_get_imports)(RZ_NONNULL ELFOBJ *bin);
RZ_BORROW RzBinElfSymbol *Elf_(rz_bin_elf_get_symbols)(RZ_NONNULL ELFOBJ *bin);

# elf_segments.c

# elf_information - where all these is_executable, is_nx, etc go

bool Elf_(rz_bin_elf_has_nx)(RZ_NONNULL ELFOBJ *bin);
bool Elf_(rz_bin_elf_has_va)(ELFOBJ *bin);
bool Elf_(rz_bin_elf_is_executable)(RZ_NONNULL ELFOBJ *bin);
bool Elf_(rz_bin_elf_is_relocatable)(RZ_NONNULL ELFOBJ *bin);
bool Elf_(rz_bin_elf_is_static)(RZ_NONNULL ELFOBJ *bin);

# elf_relocs.c

RZ_BORROW RzBinElfReloc *Elf_(rz_bin_elf_get_relocs)(RZ_NONNULL ELFOBJ *bin);

# elf_notes.c

# elf_core.c

ut64 Elf_(rz_bin_elf_get_sp_val)(RZ_NONNULL ELFOBJ *bin);

RZ_IPI RZ_BORROW RzBinElfNotePrStatus *Elf_(rz_bin_elf_get_prstatus)(RZ_NONNULL ELFOBJ *bin);
RZ_IPI RZ_BORROW RzBinElfPrStatusLayout *Elf_(rz_bin_elf_get_prstatus_layout)(RZ_NONNULL ELFOBJ *bin);

# ???
RZ_IPI Elf_(Verdaux) Elf_(rz_bin_elf_get_verdaux_entry)(RZ_NONNULL ELFOBJ *bin, ut64 offset);
RZ_IPI Elf_(Verdef) Elf_(rz_bin_elf_get_verdef_entry)(RZ_NONNULL ELFOBJ *bin, ut64 offset);
RZ_IPI Elf_(Vernaux) Elf_(rz_bin_elf_get_vernaux_entry)(RZ_NONNULL ELFOBJ *bin, ut64 offset);
RZ_IPI Elf_(Verneed) Elf_(rz_bin_elf_get_verneed_entry)(RZ_NONNULL ELFOBJ *bin, ut64 offset);
RZ_IPI RZ_OWN Sdb *Elf_(rz_bin_elf_get_version_info)(RZ_NONNULL ELFOBJ *bin);
RZ_IPI RZ_OWN Sdb *Elf_(rz_bin_elf_get_version_info_gnu_verdef)(RZ_NONNULL ELFOBJ *bin);
RZ_IPI RZ_OWN Sdb *Elf_(rz_bin_elf_get_version_info_gnu_verneed)(RZ_NONNULL ELFOBJ *bin);
RZ_IPI RZ_OWN Sdb *Elf_(rz_bin_elf_get_version_info_gnu_versym)(RZ_NONNULL ELFOBJ *bin);
RZ_IPI RZ_OWN char *Elf_(rz_bin_elf_get_ver_flags)(ut32 flags);
RZ_IPI bool Elf_(rz_bin_elf_init_dynamic_section)(RZ_NONNULL RZ_INOUT ELFOBJ *bin);
RZ_IPI bool Elf_(rz_bin_elf_init_dynstr)(RZ_NONNULL RZ_INOUT ELFOBJ *bin);
RZ_IPI bool Elf_(rz_bin_elf_init_ehdr)(RZ_NONNULL RZ_INOUT ELFOBJ *bin);
RZ_IPI bool Elf_(rz_bin_elf_init_notes)(RZ_NONNULL RZ_INOUT ELFOBJ *bin);
RZ_IPI bool Elf_(rz_bin_elf_init_phdr)(RZ_NONNULL RZ_INOUT ELFOBJ *bin);
RZ_IPI bool Elf_(rz_bin_elf_init_shdr)(RZ_NONNULL RZ_INOUT ELFOBJ *bin);
RZ_IPI bool Elf_(rz_bin_elf_init_shstrtab)(RZ_NONNULL RZ_INOUT ELFOBJ *bin);
RZ_IPI bool Elf_(rz_bin_elf_init_strtab)(RZ_NONNULL RZ_INOUT ELFOBJ *bin);


RZ_BORROW const ut8 *Elf_(rz_bin_elf_grab_regstate)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL size_t *size);
RZ_OWN ELFOBJ *Elf_(rz_bin_elf_new_buf)(RZ_NONNULL RzBuffer *buf, bool verbose);
RZ_OWN RzBinElfField *Elf_(rz_bin_elf_get_fields)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN RzBinElfLib *Elf_(rz_bin_elf_get_libs)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN RzBinElfSection *Elf_(rz_bin_elf_get_sections)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN RzBinImport *Elf_(rz_bin_elf_convert_import)(RZ_UNUSED ELFOBJ *bin, RZ_NONNULL RzBinElfSymbol *symbol);
RZ_OWN RzBinSymbol *Elf_(rz_bin_elf_convert_symbol)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL RzBinElfSymbol *symbol, const char *namefmt);
RZ_OWN RzList *Elf_(section_flag_to_rzlist)(ut64 flag);
RZ_OWN char *Elf_(rz_bin_elf_compiler)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_abi)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_arch)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_cpu)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_data_encoding)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_elf_class)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_file_type)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_head_flag)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_machine_name)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_osabi_name)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_rpath)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_intrp)(RZ_NONNULL ELFOBJ *bin);

bool Elf_(rz_bin_elf_get_stripped)(RZ_NONNULL ELFOBJ *bin);

int Elf_(rz_bin_elf_get_bits)(RZ_NONNULL ELFOBJ *bin);
int Elf_(rz_bin_elf_has_relro)(RZ_NONNULL ELFOBJ *bin);
int Elf_(rz_bin_elf_is_big_endian)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_baddr)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_boffset)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_entry_offset)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_fini_offset)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_init_offset)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_main_offset)(RZ_NONNULL ELFOBJ *bin);


ut64 Elf_(rz_bin_elf_p2v_new)(RZ_NONNULL ELFOBJ *bin, ut64 paddr);
ut64 Elf_(rz_bin_elf_v2p_new)(RZ_NONNULL ELFOBJ *bin, ut64 vaddr);
void Elf_(rz_bin_elf_free)(RZ_NONNULL ELFOBJ *bin);
```

**Comment by ghost** [comment]
I start sorting functions in different category, but some function are hard to categorize.

**Comment by ghost** [comment]
Should I add RZ_API ? (I used RZ_IPI for internal function)

**Comment by thestr4ng3r** [comment]
Some proposals (not exhaustive):
```
# elf_core.c
RZ_BORROW const ut8 *Elf_(rz_bin_elf_grab_regstate)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL size_t *size);

# elf.c

RZ_OWN ELFOBJ *Elf_(rz_bin_elf_new_buf)(RZ_NONNULL RzBuffer *buf, bool verbose);
void Elf_(rz_bin_elf_free)(RZ_NONNULL ELFOBJ *bin);

# elf_info.c

RZ_OWN char *Elf_(rz_bin_elf_compiler)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_abi)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_arch)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_cpu)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_data_encoding)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_elf_class)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_file_type)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_head_flag)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_machine_name)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_osabi_name)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_rpath)(RZ_NONNULL ELFOBJ *bin);
bool Elf_(rz_bin_elf_get_stripped)(RZ_NONNULL ELFOBJ *bin);
int Elf_(rz_bin_elf_get_bits)(RZ_NONNULL ELFOBJ *bin);
int Elf_(rz_bin_elf_has_relro)(RZ_NONNULL ELFOBJ *bin);
int Elf_(rz_bin_elf_is_big_endian)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_baddr)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_boffset)(RZ_NONNULL ELFOBJ *bin);

# elf_symbols.c

RZ_OWN RzBinImport *Elf_(rz_bin_elf_convert_import)(RZ_UNUSED ELFOBJ *bin, RZ_NONNULL RzBinElfSymbol *symbol);
RZ_OWN RzBinSymbol *Elf_(rz_bin_elf_convert_symbol)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL RzBinElfSymbol *symbol, const char *namefmt);
```

We should probably have all of these `Elf_` functions with the same `RZ_XPI`. I would say `RZ_API` since they are getting more and more into a shape where they can also be used by themselves outside the plugin, or for getting elf-specific info when using RzBin as a C library. @XVilka @ret2libc @wargio what do you think?

**Comment by XVilka** [comment]
I have no strong opinion about this, but probably isn't very consistent with the spirit of the framework to provide unified API for all formats/architectures.

**Comment by XVilka** [comment]
The functions that are hard to categorize are a perfect fit for the "misc" category.

**Comment by ghost** [comment]
> The functions that are hard to categorize are a perfect fit for the "misc" category.

If a function is not used by the plugin, it should be removed? (because it isn´t a API)

**Comment by XVilka** [comment]
@08A yes, if it has no place in your future refactoring for improving ELF loading - remove it.

**Comment by ghost** [comment]
// elf_deprecated.c

RZ_DEPRECATE bool Elf_(rz_bin_elf_del_rpath)(RzBinFile *bf);
RZ_DEPRECATE bool Elf_(rz_bin_elf_entry_write)(RzBinFile *bf, ut64 addr);
RZ_DEPRECATE bool Elf_(rz_bin_elf_section_perms)(RzBinFile *bf, const char *name, int perms);
RZ_DEPRECATE ut64 Elf_(rz_bin_elf_p2v)(RZ_NONNULL ELFOBJ *bin, ut64 paddr);
RZ_DEPRECATE ut64 Elf_(rz_bin_elf_resize_section)(RzBinFile *bf, const char *name, ut64 size);
RZ_DEPRECATE ut64 Elf_(rz_bin_elf_v2p)(RZ_NONNULL ELFOBJ *bin, ut64 vaddr);

// elf_misc.c

ut64 Elf_(rz_bin_elf_p2v_new)(RZ_NONNULL ELFOBJ *bin, ut64 paddr);
ut64 Elf_(rz_bin_elf_v2p_new)(RZ_NONNULL ELFOBJ *bin, ut64 vaddr);

// elf.c

RZ_OWN ELFOBJ *Elf_(rz_bin_elf_new_buf)(RZ_NONNULL RzBuffer *buf, bool verbose);
void Elf_(rz_bin_elf_free)(RZ_NONNULL ELFOBJ *bin);

// elf_core.c

RZ_BORROW const ut8 *Elf_(rz_bin_elf_grab_regstate)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL size_t *size);
ut64 Elf_(rz_bin_elf_get_sp_val)(RZ_NONNULL ELFOBJ *bin);

// elf_info.c

RZ_OWN RzBinElfField *Elf_(rz_bin_elf_get_fields)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN RzBinElfLib *Elf_(rz_bin_elf_get_libs)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_compiler)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_abi)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_arch)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_cpu)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_data_encoding)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_elf_class)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_file_type)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_head_flag)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_machine_name)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_osabi_name)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_get_rpath)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN char *Elf_(rz_bin_elf_intrp)(RZ_NONNULL ELFOBJ *bin);
bool Elf_(rz_bin_elf_get_stripped)(RZ_NONNULL ELFOBJ *bin);
bool Elf_(rz_bin_elf_has_nx)(RZ_NONNULL ELFOBJ *bin);
bool Elf_(rz_bin_elf_has_va)(ELFOBJ *bin);
bool Elf_(rz_bin_elf_is_executable)(RZ_NONNULL ELFOBJ *bin);
bool Elf_(rz_bin_elf_is_relocatable)(RZ_NONNULL ELFOBJ *bin);
bool Elf_(rz_bin_elf_is_static)(RZ_NONNULL ELFOBJ *bin);
int Elf_(rz_bin_elf_get_bits)(RZ_NONNULL ELFOBJ *bin);
int Elf_(rz_bin_elf_has_relro)(RZ_NONNULL ELFOBJ *bin);
int Elf_(rz_bin_elf_is_big_endian)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_baddr)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_boffset)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_entry_offset)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_fini_offset)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_init_offset)(RZ_NONNULL ELFOBJ *bin);
ut64 Elf_(rz_bin_elf_get_main_offset)(RZ_NONNULL ELFOBJ *bin);

// elf_notes.c

// elf_relocs.c

RZ_BORROW RzBinElfReloc *Elf_(rz_bin_elf_get_relocs)(RZ_NONNULL ELFOBJ *bin);

// elf_sections.c

RZ_BORROW RzBinElfSection *Elf_(rz_bin_elf_get_section)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL const char *section_name);
RZ_OWN RzBinElfSection *Elf_(rz_bin_elf_get_sections)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN RzList *Elf_(section_flag_to_rzlist)(ut64 flag);
RZ_OWN char *Elf_(section_type_to_string)(ut64 type);
ut64 Elf_(rz_bin_elf_get_section_addr)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL const char *section_name);
ut64 Elf_(rz_bin_elf_get_section_addr_end)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL const char *section_name);
ut64 Elf_(rz_bin_elf_get_section_offset)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL const char *section_name);

// elf_segment.c

// elf_symbols.c

RZ_BORROW RzBinElfSymbol *Elf_(rz_bin_elf_get_imports)(RZ_NONNULL ELFOBJ *bin);
RZ_BORROW RzBinElfSymbol *Elf_(rz_bin_elf_get_symbols)(RZ_NONNULL ELFOBJ *bin);
RZ_OWN RzBinImport *Elf_(rz_bin_elf_convert_import)(RZ_UNUSED ELFOBJ *bin, RZ_NONNULL RzBinElfSymbol *symbol);
RZ_OWN RzBinSymbol *Elf_(rz_bin_elf_convert_symbol)(RZ_NONNULL ELFOBJ *bin, RZ_NONNULL RzBinElfSymbol *symbol, const char *namefmt);

**Comment by ghost** [comment]
This configuration is better?

Can I renamed some function like `rz_bin_elf_compiler` to `rz_bin_elf_get_compiler`

**Comment by XVilka** [comment]
```
// elf_misc.c

ut64 Elf_(rz_bin_elf_p2v_new)(RZ_NONNULL ELFOBJ *bin, ut64 paddr);
ut64 Elf_(rz_bin_elf_v2p_new)(RZ_NONNULL ELFOBJ *bin, ut64 vaddr);
``` 
I think, `v2p` and `p2v` functions should be either in `elf.c` or in `elf_core.c`, since they are vital, IMHO.

**Comment by XVilka** [comment]
Also,
```
RZ_DEPRECATE ut64 Elf_(rz_bin_elf_p2v)(RZ_NONNULL ELFOBJ *bin, ut64 paddr);
RZ_DEPRECATE ut64 Elf_(rz_bin_elf_v2p)(RZ_NONNULL ELFOBJ *bin, ut64 vaddr);
```
Are they still used anywhere? What about changing the API instead of adding two new functions with `_new` suffix?

**Comment by ghost** [comment]
> ```
> // elf_misc.c
> 
> ut64 Elf_(rz_bin_elf_p2v_new)(RZ_NONNULL ELFOBJ *bin, ut64 paddr);
> ut64 Elf_(rz_bin_elf_v2p_new)(RZ_NONNULL ELFOBJ *bin, ut64 vaddr);
> ```
> 
> I think, `v2p` and `p2v` functions should be either in `elf.c` or in `elf_core.c`, since they are vital, IMHO.


elf_core is about core file, I will move them to elf.c

**Comment by ghost** [comment]
> Also,
> 
> ```
> RZ_DEPRECATE ut64 Elf_(rz_bin_elf_p2v)(RZ_NONNULL ELFOBJ *bin, ut64 paddr);
> RZ_DEPRECATE ut64 Elf_(rz_bin_elf_v2p)(RZ_NONNULL ELFOBJ *bin, ut64 vaddr);
> ```
> 
> Are they still used anywhere? What about changing the API instead of adding two new functions with `_new` suffix?

All RZ_DEPRECATE function are unused. If you want I can removed them.

**Comment by XVilka** [comment]
Just remove them then, and drop `_new` prefix for `v2p` and `p2v`. Also maybe rename `elf_core.c` to `elf_corefile.c` to be more clear about purpose.

**Comment by XVilka** [comment]
> Can I renamed some function like `rz_bin_elf_compiler` to `rz_bin_elf_get_compiler`

Yes, lets keep our naming consistent.

**Comment by ghost** [comment]
* merged some files
* undeprecated elf_write.c (It was used...)
* harmonized the naming convention
* delete v2p, p2v and rename v2p_new, p2v_new

**Comment by ghost** [comment]
**Warning** get_symbols_with_type hasn't been refactor (spend to much time on this)

File: librz/bin/format/elf/elf.c [hunk 2]
```diff
@@ -712,3604 +687,453 @@ static void fill_dynamic_entries(ELFOBJ *bin, ut64 loaded_offset, ut64 dyn_size)
...
-		return;
-	}
-	if (seg->notes) {
-		for (size_t i = 0; i < seg->notes_count; i++) {
-			note_fini(&seg->notes[i]);
-		}
-		free(seg->notes);
-	}
-	free(seg);
+static void set_note_file(ELFOBJ *bin, RzBinElfNoteFile *file, ut64 *offset, const char *file_name) {
```
Comment (thestr4ng3r) [inline librz/bin/format/elf/elf.c]: All of these notes, at least the ones we parse, are specific to core files, so they can go into `elf_corefile.c`.

**Review by thestr4ng3r — CHANGES_REQUESTED** [review]
> https://soundcloud.com/powerwolf-official/beast-of-gevaudan

Good choice!

> undeprecated elf_write.c (It was used...)

deprecated in Rizin does not mean something is unused but only that something should not be used in new code. If something is deprecated and unused, it should be removed directly.

Overall looks pretty good to me now.

File: librz/bin/format/elf/elf.c (hunk 2 above)
Comment (ghost) [reply librz/bin/format/elf/elf.c]: I can add (tomorrow) a elf_notes.c file if you want.

Comment (thestr4ng3r) [reply librz/bin/format/elf/elf.c]: sounds good to me.

**Review by thestr4ng3r — DISMISSED** [review]

**Review by thestr4ng3r — APPROVED** [review]
Good job!

File: librz/bin/format/elf/elf.h [hunk 3]
```diff
@@ -12,32 +12,93 @@
...
+	(i) += 1
+#define BREAD16(x, i) \
+	rz_buf_read_ble16_at(x, i, bin->endian); \
+	(i) += 2
+#define BREAD32(x, i) \
+	rz_buf_read_ble32_at(x, i, bin->endian); \
+	(i) += 4
+#define BREAD64(x, i) \
+	rz_buf_read_ble64_at(x, i, bin->endian); \
+	(i) += 8
```
Comment (wargio) [inline librz/bin/format/elf/elf.h]: tbh, i would prefer to see `(i) += sizeof(utXX)`

File: librz/bin/format/elf/elf_sections.c [hunk 4]
```diff
@@ -0,0 +1,311 @@
+// SPDX-FileCopyrightText: 2008-2020 nibble <nibble.ds@gmail.com>
+// SPDX-FileCopyrightText: 2008-2020 pancake <pancake@nopcode.org>
+// SPDX-FileCopyrightText: 2008-2020 alvaro_fe <alvaro.felipe91@gmail.com>
```
Comment (wargio) [inline librz/bin/format/elf/elf_sections.c:4]: Add your name here.

File: test/db/cmd/cmd_i [hunk 5]
```diff
@@ -4138,7 +4138,7 @@ vaddr      paddr      target     type   name
 0x00001ff0 4 reloc.__gmon_start
 0x00001ff8 4 reloc._Jv_RegisterClasses
 0x00001ffc 4 reloc._ITM_registerTMCloneTable
-0x00002000 16 section..got.plt
+0x00002000 0 section..got.plt
```
Comment (wargio) [inline test/db/cmd/cmd_i:4141]: why this change? is this a bug

File: test/db/tools/rz_bin [hunk 6]
```diff
@@ -678,19 +678,19 @@ vaddr      paddr      type   name
...
 
 
-Version need section '.gnu.version_r' contains 1 entries:
- Addr: 0x08048250  Offset: 0x00000250  Link to section: 6 (.dynstr)
-  0x00000000: Version: 1  File: libc.so.6  Cnt: 1
-  0x00000010:   Name: GLIBC_2.0  Flags: none Version: 2
+Version need has 1 entries:
+ Addr: 0x08048250  Offset: 0x00000250
+  0x00000250: Version: 1  File: libc.so.6  Cnt: 1
+  0x00000260:   Name: GLIBC_2.0  Flags: none Version: 2
```
Comment (wargio) [inline test/db/tools/rz_bin:693]: why these changes of offset?

**Review by wargio — CHANGES_REQUESTED** [review]

File: librz/bin/format/elf/elf_sections.c (hunk 4 above)
Comment (wargio) [reply librz/bin/format/elf/elf_sections.c:4]: also in the other files

File: test/db/cmd/cmd_i (hunk 5 above)
Comment (ghost) [reply test/db/cmd/cmd_i:4141]: When there is no section information, sections are created with DT_DYNAMIC information. However there is no infomation about the .plt.got length. So i choose 0.

File: test/db/tools/rz_bin (hunk 6 above)
Comment (ghost) [reply test/db/tools/rz_bin:693]: The old offset was the offset from the .gnu.version_r. I switch to the offset in the file.

The reason is that it simplify the implementation.

Comment (ghost) [reply test/db/tools/rz_bin:693]: In the example the first offset is the verneed struct and the second offset is a vernaux struct.

File: test/db/cmd/cmd_i (hunk 5 above)
Comment (wargio) [reply test/db/cmd/cmd_i:4141]: makes sense

**Review by wargio — APPROVED** [review]

**Review by XVilka — APPROVED** [review]
